src/base/autoload.lisp

```diff
@@ -18,14 +19,47 @@
                         (lambda (&rest ,args)
                           (declare (ignore ,args))
                           (error "Autoloading ~S failed." ',name)))
+
                   (when (asdf:load-system ',asdf-system-name)
-                    (setf (box-value ,box) ,actual-body))
-                  (funcall (box-value ,box) ,box t))))
+                    ;; If the system loads correctly, set the function
+                    ;; within BOX to a trampoline to the actual body of
+                    ;; the WITH-LOADING form. Note that we cannot simply
+                    ;; store the current value of ACTUAL-BODY within BOX
+                    ;; because that would mean the lexical scope of BODY
+                    ;; would always correspond to the outer scope at the
+                    ;; time of the *first* invocation, and not any
+                    ;; subsequent ones.
+                    (setf (box-value ,box)
+                          (lambda (,box ,first-time-p ,actual-body)
+                            (declare (ignore ,box))
+                            (funcall ,actual-body ,first-time-p))))
+
+                  ;; Finally, call the actual body, as appropriate.
+                  (funcall (box-value ,box) ,box t ,actual-body))))
+    ;; At load time, we create a BOX object containing a function
+    ;; corresponding to the body to be executed by this WITH-AUTOLOADING
+    ;; form. The initial value is a function that attempts to load the
+    ;; ASDF system ASDF-SYSTEM-NAME, setting the value within BOX to
+    ;; either a function signalling that autoloading failed, or a
+    ;; trampoline to the provided BODY, depending on whether the system
+    ;; loaded successfully or not. Further executions will always invoke
+    ;; either the signalling function or the trampoline, without
+    ;; attempting to load the ASDF system again.
```

Comment:
Ah, I didn't realize that the trampoline remains after autoloading. I would prefer if autoloading and performance were entirely orthogonal consideration. I mean, `WITH-AUTOLOADING` is fine, but I wouldn't implement `AUTOLOAD` with it. Does that sound reasonable?

Reply:
> Ah, I didn't realize that the trampoline remains after autoloading.

It won't in the case of `autoload`, because we still expect the systems of `autoload`ed functions to replace their `symbol-function` with the actual function body (if it does remain, it will be the one calling the body that signals an error telling the user that the given system has loaded successfully, but failed to redefine the current function).

Reply:
Of course, thanks.